src/csharp/Microsoft.Spark/Utils/UdfSerDe.cs

```diff
@@ -54,21 +62,175 @@ internal sealed class UdfData
             public TypeData TypeData { get; set; }
             public string MethodName { get; set; }
             public TargetData TargetData { get; set; }
+
+            public override int GetHashCode()
+            {
+                return base.GetHashCode();
+            }
+
+            public override bool Equals(object obj)
+            {
+                return (obj is UdfData udfData) &&
+                    Equals(udfData);
+            }
+
+            public bool Equals(UdfData other)
+            {
+                return (other != null) &&
+                    (TypeData.Equals(other.TypeData)) &&
+                    (MethodName == other.MethodName) &&
+                    (TargetData.Equals(other.TargetData));
+            }
         }
 
         [Serializable]
         internal sealed class TargetData
         {
             public TypeData TypeData { get; set; }
             public FieldData[] Fields { get; set; }
+
+            public override int GetHashCode()
+            {
+                return base.GetHashCode();
+            }
+
+            public override bool Equals(object obj)
+            {
+                return (obj is TargetData targetData) &&
+                    Equals(targetData);
+            }
+
+            public bool Equals(TargetData other)
+            {
+                if ((other == null) ||
+                    !TypeData.Equals(other.TypeData) ||
+                    (Fields?.Length != other.Fields?.Length))
+                {
+                    return false;
+                }
+                else if (Fields == null && other.Fields == null)
```

Comment:
shouldn't this by just `if`?

Reply:
also, according to https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions#layout-conventions
it should be:
```C#
if ((Fields == null) && (other.Fields == null))
```
(check other places as well)